Add guidelines to deal with user permissions for specific properties in objects (#333)
teamleadercrm/api: GUIDELINES.md

 - `204` - No Content, on resource updates or actions
 - `400` - Bad Request, the request contains invalid data or references non-existing resources
 - `401` - Unauthorized, invalid or missing access token

>  - `403` - Forbidden, not allowed to access this resource, or a specific property of a resource

Author: We thought about returning a `400` when the User don't have permissions to access a specific property but it would be incorrect because the request is not malformed nor invalid, is indeed a matter of authorization.

Reviewer: When reading this, this would also return a 403 when retrieving info about a resource with a property you don't have access to. Would change it to:

```suggestion
 - `403` - Forbidden, not allowed to access this resource or update a property you are not allowed to change
```

}
```

### Restricted properties

> When a User don't have permissions to access a specific property the endpoint **MUST**:

Reviewer 2: ```suggestion
When a User doesn't have permissions to access a specific property the endpoint **MUST**:
```

 - Not return that **property** in the response.

>  - When updating or creating an Object and it contains that **property**, then the endpoint **MUST** return a `403` error response with an appropriate explanation. The property **MUST** be optional.

Reviewer 3: ```suggestion
 - When writing to a resource that contains that **property**, the endpoint **MUST** return a `403` error response with an appropriate message.
```

}
```

> ### User Permissions

Reviewer: This could be under a section called:

```suggestion
### Restricted properties
```

#### Fetching objects

> When fetching objects if the User has permissions to fetch it but he lacks a specific property permission then that property will not be returned in the response.

Author: We discussed the possibility to simple return `null` instead of making the properties `optional` in the response. If we would make them `null` that would mean that the integrator wouldn't know if the property doesn't have value or if the User don't have the right to access it. Making it optional will make it explicit that the User don't have access to it but will make it less friendly for our integrators because they will have to have some if conditions for those properties.

> When fetching objects if the User has permissions to fetch it but he lacks a specific property permission then that property will not be returned in the response.

Reviewer: Would also not split up the reading and writing since they're both related to restricted properties.

```suggestion
Sometimes users don't have the required permissions to read or write certain data. In our API responses, we should not return the properties the user has no access to.

If that user tries to change a property he is not allowed to change, then we will return a `403` error response with an appropriate explanation.
```

> When fetching objects if the User has permissions to fetch it but he lacks a specific property permission then that property will not be returned in the response.

Reviewer 2: ```suggestion
When fetching objects, if the User has permissions to fetch but lacks a specific property permission, that property will not be returned in the response.
```

#### Creating/Updating objects

> A User that has permission to add a Object might not have permissions to certain properties of that object. In the case 

Reviewer 2: ```suggestion
A User that has permission to add an Object might not have permissions to certain properties of that object. In case 
```

> he tries to add or update a object with a property that he does not have access to, the request will fail and will 

Reviewer 2: ```suggestion
they try to add or update an Object with a property that they don't have access to, the request will fail and 
```

> return a `403` response with a message explaining what caused it.

Reviewer 2: ```suggestion
a `403` response with an explanatory message will be returned.
```
